api/src/wfl/util.clj

```diff
@@ -432,3 +434,81 @@
    (poll task seconds 3))
   ([task]
    (poll task 1)))
+
+(defn columns-rows->tsv
+  "
+  Writes COLUMNS and ROWS to a .tsv named FILE, or to bytes if FILE not specified.
+
+  Examples
+  --------
+    (columns-rows->tsv [c1 c2 c3] [[x1 x2 x3] [y1 y2 y3] ...])
+    (columns-rows->tsv [c1 c2 c3] [[x1 x2 x3] [y1 y2 y3] ...] \"destination.tsv\")
+  "
+  ([columns rows file]
+   (with-open [writer (io/writer file)]
+     (csv/write-csv writer columns :separator \tab)
+     (csv/write-csv writer rows :separator \tab)
+     file))
+  ([columns rows]
+   (str (columns-rows->tsv columns rows (StringWriter.)))))
+
+"A set of all mutually supported TSV file types (by FireCloud and WFL)"
+(s/def ::tsv-type #{:entity :membership})
+
+(defn terra-id
+  "
+  Generates a Terra-compatible primary key column name for the specified TSV-TYPE and base COL.
+
+  The first column of the table must be its primary key, and named accordingly:
+    entity:[entity-type]_id
+    membership:[entity-type]_set_id
+
+  For tsv-type :entity, 'entity:sample_id' will upload the .tsv data into a `sample` table
+  in the workspace (or create one if it does not exist).
+  If the table already contains a sample with that id, it will get overwritten.
+
+  For tsv-type :membership, 'membership:sample_set_id' will append sample names to a `sample_set` table
+  in the workspace (or create one if it does not exist).
+  If the table already contains a sample set with that id, it will be appended to and not overwritten.
```

Comment:
Flagging this: for entity set uploads (`tsv-type :membership`) the FireCloud behavior for existing ids doesn't match that for entity uploads (`tsv-type :entity`).

Reply:
I've since observed that while an entity set shows duplicate entry references, if you create a submission off of the entity set it will only create one workflow for each distinct entity reference.